---
title: "HMM analyses"
author: "LJM"
date: "`r Sys.Date()`"
link-citations: true
bibliography: bibliography.bib
biblio-style: "alpha" #https://www.overleaf.com/learn/latex/Bibtex_bibliography_styles
#csl: "../uppsala-universitet-institutionen-for-biologisk-grundutbildning.csl"
documentclass: report
output:
  bookdown::html_document2:
    highlight: tango
    theme: journal
    split_by: none # only generate a single output page
    self_contained: TRUE
    toc: yes
    toc_float: 
      collapsed: TRUE
      smooth_scroll: TRUE
      print: FALSE
    code_folding: show
    number_sections: TRUE
    code_download: TRUE
  html_document: 
    toc: TRUE
    toc_float: TRUE
    code_folding: show
    highlight: tango
    number_sections: TRUE
    self_contained: TRUE
    smart: TRUE # To be figured out when this is useful
    theme: journal
  pdf_document:
    toc: TRUE
    toc_depth: 2
    highlight: tango # This appears to be the same as default
    number_sections: TRUE
  bookdown::pdf_document2:
    keep_tex: no
    latex_engine: xelatex
    #highlight: pygments #Other options: default, tango, kate, monochrome, espresso, zenburn, haddock, and textmate
    #theme: spacelab
    pandoc_args: ["--variable", "subparagraph",
                "--top-level-division","chapter"] # Could also be section 
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Find annotated viruses in the `.emapper.annotations` files

Search all viruses from eggnogg mapper annotations and store them in `viruses.emapper.annotations`-file. The goal is to use the identifiers in this file to fetch out protein sequences of all these viruses. 

One possible way to go might be to create profile HMMs of some clear groups of viruses, e.g. all _Myoviridae_ are used to create a _Myoviridae_ profile, all _Siphoviridae_ another profile, etc. and scan the fOTUs with these HMM profiles to classify them. 

## Gather all eggnogg mapper annotated viruses in one file

```{bash eval=FALSE, include=TRUE}
ANNOT="../data/annotations/"
ANALYSES="../analyses/HMM_scan/"

grep -E 'Viruses' "$ANNOT"*.emapper.annotations > "$ANALYSES"viruses.emapper.annot
head -n 2 "$ANALYSES"viruses.emapper.annot
```

## Survey what categories of viruses are there

Now that we have all the viruses found, it would be good to try to survey what type of categories of viruses are there so that they can be used to create the profiles. Maybe one way to see the categories is the following:

```{bash eval=FALSE, include=TRUE}
ANNOT="../data/annotations/"
ANALYSES="../analyses/HMM_scan/"

awk '{print $5}' "$ANALYSES"viruses.emapper.annot | sort | uniq
```

We thus see that there are 9 different categories of viruses with varying specificity. 

## Gather and group all viral protein sequences

Next task is to gather groupwise all viral protein sequences. Maybe one simple way to go about doing that is by pattern matching the 5th field in the previously produced `viruses.emapper.annot` file with the found categories and saving each category in own file.

```{bash eval=FALSE, include=TRUE}
INPUT="../analyses/HMM_scan/viruses.emapper.annot"
SCRIPT="../scripts/virusSeparator.awk"
RESULTS_DIR="../analyses/HMM_scan/virus_annotations_categorised/"
cat "$INPUT" | awk -v output_dir="$RESULTS_DIR" -f "$SCRIPT"
```

This seems to be possible by searching the correct file indexing it with exonerate v. 2.4.0 [@Slater2005] command `fastaindex` and then fetching it with another tool in the exonerate utility suite called `fastafetch`. Lastly each of the fetched files are appended to a new file with hopefully most of the annotation data not lost (but contained in the fasta header).


```{bash eval=FALSE, include=TRUE}
VIRUS_ANNOT="../analyses/HMM_scan/virus_annotations_categorised/"
PROTEOMS="../data/proteoms/"
SCRIPT="../scripts/virusProteomGatherer.awk"
PROTEOM_INDEXS="../analyses/HMM_scan/proteom_indexes/"
VIRUS_PROTEOMS="../analyses/HMM_scan/virus_proteoms/"

read -r -a ANNOTs <<< $( find $VIRUS_ANNOT -name "*.annot" -and -type f -print0 | xargs -0 echo )
for ANNOT in "${ANNOTs[@]}"; do
  FILE=$(echo $(basename "$ANNOT")) 
  VIR_CATEGORY=$(echo $(basename "$ANNOT") | awk -F "." '{print $1}')
  #echo "$FILE $VIR_CATEGORY"
  cat "$VIRUS_ANNOT""$FILE" | awk -v output_dir="$RESULTS_DIR" -v proteom_dir="$PROTEOMS" -v proteom_index_dir="$PROTEOM_INDEXS" -f "$SCRIPT" > "$VIRUS_PROTEOMS""$VIR_CATEGORY"".faa"
done

```


The tool to be used in the HMM analyses is called HMMer v. 3.2.1 [@hmmer2018].

```{bash eval=FALSE, include=FALSE}
fastaindex --help
#fastafetch --help
#fastaindex --fasta "../data/proteoms/Loc090116-9m_megahit_metabat_bin-1153.faa" --index "../analyses/HMM_scan/proteom_indexes/Loc090116-9m_megahit_metabat_bin-1153.idx"
#cut -f 2- "$ANALYSES"viruses.emapper.annotations | head -n 3

#fastafetch --fasta "../data/proteoms/Loc090116-9m_megahit_metabat_bin-1153.faa" --index "../analyses/HMM_scan/proteom_indexes/Loc090116-9m_megahit_metabat_bin-1153.idx" --query "Loc090116-9m_megahit_metabat_bin-1153_00054"

```



# References
